protocol/contracts/nexus/Shard.sol

```diff
@@ -0,0 +1,555 @@
+            } else if (!allowed && partnerShardIds.contains(shardId)){
+                /* success = */ partnerShardIds.remove(shardId);
+                if (partnerShards.contains(shardId)) {
+                    partnerShards.remove(shardId);
+                }
+            }
+            unchecked {
+                 ++i;
+             }
+         }
+         emit PartnerAllowedShardIdsSet(partner, shardIds, flags);
+    }
+
+    /*
+     * Set the caps for shards partners can mint
+     * @param partner Address of the partner
+     * @param shardIds Shard IDs
+     * @param caps The maximum amount partner can mint for each shard
+     */
+    function setPartnerAllowedShardCaps(
+        address partner,
+        uint256[] memory shardIds,
+        uint256[] memory caps
+    ) external onlyElevatedAccess {
+        /// @notice 0 by default which is unlimited
+        if (partner == ZERO_ADDRESS) { revert ZeroAddress(); }
+        uint256 _length = shardIds.length;
+        if (_length != caps.length) { revert InvalidParamLength(); }
+        for (uint i; i < _length; ) {
+            allowedPartnersShardCaps[partner][shardIds[i]] = caps[i];
+            unchecked {
+                ++i;
+            }
+        }
+        emit PartnerAllowedShardCapsSet(partner, shardIds, caps);
+    }
+
+    /*
+     * Set a recipe for transmutation
+     * @param recipeId The recipe ID
+     * @param recipe The recipe
+     */
+    function setRecipe(uint256 recipeId, Recipe calldata recipe) external onlyElevatedAccess {
+        uint256 _inputLength = recipe.inputShardIds.length;
+        if (_inputLength != recipe.inputShardAmounts.length) { revert CommonEventsAndErrors.InvalidParam(); }
+        uint256 _outputLength = recipe.outputShardAmounts.length;
+        if (_outputLength != recipe.outputShardIds.length) { revert CommonEventsAndErrors.InvalidParam(); }
+
+        recipes[recipeId] = recipe;
+        emit RecipeSet(recipeId, recipe);
+    }
+
+    /*
+     * Delete recipe
+     * @param recipeId The recipe ID
+     */
+    function deleteRecipe(uint256 recipeId) external onlyElevatedAccess {
+        Recipe storage recipe = recipes[recipeId];
+        if (recipe.inputShardIds.length == 0) { revert CommonEventsAndErrors.InvalidParam(); }
+        delete recipes[recipeId];
+        emit RecipeDeleted(recipeId);
+    }
+
+    /*
+     * Set uri string of shard ID
+     * @param shardId The shard ID
+     * @param uri The uri string
+     * @return String uri of the shard ID
+     */
+    function setShardUri(uint256 shardId, string memory uri) external onlyElevatedAccess {
+        if (bytes(uri).length == 0 ) { revert CommonEventsAndErrors.InvalidParam(); }
+        shardUris[shardId] = uri;
+        emit ShardUriSet(shardId, uri);
+    }
+
+    /*
+     * Get uri string of shard ID
+     * @param shardId The shard ID
+     * @return String uri of the shard ID
+     */
+    function uri(uint256 shardId) public view override returns (string memory) {
```

Comment:
This is identical functionality to the auto getter `shardUris()`. Could make `shardUris` private since adding a getter on that is redundant?